src/app/spans.rs

```diff
@@ -0,0 +1,106 @@
+use crate::trace_context;
+use linkerd2_error::Error;
+use opencensus_proto::trace::v1 as oc;
+use std::collections::HashMap;
+use std::{error, fmt};
+use tokio::sync::mpsc;
+
+const SPAN_KIND_SERVER: i32 = 1;
+const SPAN_KIND_CLIENT: i32 = 2;
+
+#[derive(Clone)]
+pub struct SpanConverter {
+    kind: i32,
+    sink: mpsc::Sender<oc::Span>,
+    labels: HashMap<String, String>,
+}
+
+#[derive(Debug)]
+pub struct IdLengthError {
+    id: Vec<u8>,
+    expected_size: usize,
+    actual_size: usize,
+}
+
+impl error::Error for IdLengthError {
+    fn description(&self) -> &str {
+        "trace or span id is wrong number of bytes"
+    }
+}
+
+impl fmt::Display for IdLengthError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Id '{:?}' should have {} bytes but it has {}",
+            self.id, self.expected_size, self.actual_size
+        )
+    }
+}
+
+impl SpanConverter {
+    pub fn server(sink: mpsc::Sender<oc::Span>, labels: HashMap<String, String>) -> Self {
+        Self {
+            kind: SPAN_KIND_SERVER,
+            sink,
+            labels,
+        }
+    }
+
+    pub fn client(sink: mpsc::Sender<oc::Span>, labels: HashMap<String, String>) -> Self {
+        Self {
+            kind: SPAN_KIND_CLIENT,
+            sink,
+            labels,
+        }
+    }
+
+    fn mk_span(&self, span: trace_context::Span) -> Result<oc::Span, IdLengthError> {
+        Ok(oc::Span {
+            trace_id: into_bytes(span.trace_id, 16)?,
+            span_id: into_bytes(span.span_id, 8)?,
+            tracestate: None,
+            parent_span_id: into_bytes(span.parent_id, 8)?,
+            name: Some(truncatable(span.span_name)),
+            kind: self.kind,
+            start_time: Some(span.start.into()),
+            end_time: Some(span.end.into()),
+            attributes: None, // TODO: attributes
+            stack_trace: None,
+            time_events: None,
+            links: None,
+            status: None, // TODO: record this
+            resource: None,
+            same_process_as_parent_span: Some(self.kind == SPAN_KIND_CLIENT),
+            child_span_count: None,
```

Comment:
TIOLI: instead of having so many `None`s, you _could_ leave them empty and do
```rust
oc::Span {
    // fields...
   ...oc::Span::default()
}
```

though I'd leave the fields with `TODO`s on them.